src/software/new_geom/cubic_bezier_spline2d.cpp

```diff
@@ -0,0 +1,167 @@
+#include "software/new_geom/cubic_bezier_spline2d.h"
+
+#include <assert.h>
+
+#include <algorithm>
+
+CubicBezierSpline2d::CubicBezierSpline2d(const Point& start_point,
+                                         const Vector& start_vector,
+                                         const Point& end_point, const Vector& end_vector,
+                                         std::vector<Point> intermediate_knots)
+    : control_points(computeControlPoints(start_point, start_vector, end_point,
+                                          end_vector, intermediate_knots))
+{
+}
+
+const Point CubicBezierSpline2d::getValueAt(double t) const
+{
+    // Find which segment of this spline `t` falls on, assuming the start of
+    // the first segment is t=0, the end of the last segment is t=1, and the
+    // the knots are evenly distributed in between 0 and 1
+
+
+    const double t_constrained = std::clamp(t, 0.0, 1.0);
+
+    // Scale t from [0,1] -> [0,n] where n is the number segments
+    const double t_mapped = t_constrained * static_cast<double>(getNumSegments());
+
+    // In most cases the segment index is just the floor of the mapped t, but we
+    // also need to have extra handling for the case where t=1, in which case
+    // we need to just use the last segment
+    const size_t segment_index =
+        std::min(static_cast<size_t>(std::floor(t_mapped)), getNumSegments() - 1);
+
+    // Figure out what the t-value on the segment is going to be
+    const double t_on_segment = (t_mapped - static_cast<double>(segment_index));
+
+    // Generate the bezier curve for the segment and evaluate it at the segment t-value
+    const BezierCurve2d segment = getSegmentAtIndex(segment_index);
+    return segment.getValueAt(t_on_segment);
+}
+
+const std::vector<Point> CubicBezierSpline2d::getKnots() const
+{
+    std::vector<Point> knots;
+    for (size_t i = 0; i < control_points.size(); i += 3)
+    {
+        knots.emplace_back(control_points[i]);
+    }
+
+    return knots;
+}
+
+const std::vector<Point>& CubicBezierSpline2d::getControlPoints() const
+{
+    return control_points;
+}
+
+size_t CubicBezierSpline2d::getNumKnots() const
+{
+    return static_cast<size_t>((control_points.size() - 1) / 3 + 1);
+}
+
+size_t CubicBezierSpline2d::getNumSegments() const
+{
+    return getNumKnots() - 1;
+}
+
+
+std::vector<double> CubicBezierSpline2d::getKnotParametrizationValues() const
+{
+    // We assume a linear spacing of all the knots from 0 to 1
+    std::vector<double> knot_vector;
+    const double num_knots = getNumKnots();
+    for (size_t i = 0; i < num_knots; i++)
+    {
+        knot_vector.emplace_back(i * 1.0 / static_cast<double>(num_knots - 1));
+    }
+    return knot_vector;
+}
+
+const Point CubicBezierSpline2d::getStartPoint() const
+{
+    // We are guaranteed to at least have two control points, the start and end points
```

Comment:
Are we not guaranteed 4 control points because of the initially provided vectors? Does that break down if they are zero-length vectors?

If we are only guaranteed 2, that means `getNumKnots` could return `1` which results in division by zero in `getKnotParametrizationValues`

Reply:
yep changed here and above:
```
    // We are guaranteed to at least have four control points, the start and end points,
    // along with `start + start_vector` and `end + end_vector`
```